workers/unity/Packages/com.improbable.gdk.tools/.CodeGenerator/GdkCodeGenerator/Templates/UnityComponentDataGenerator.tt

```diff
@@ -24,8 +24,81 @@ namespace <#= qualifiedNamespace #>
         {
             public uint ComponentId => <#= componentDetails.ComponentId #>;
 
-            public BlittableBool DirtyBit { get; set; }
-<# foreach(var fieldDetails in fieldDetailsList) { #>
+            private BlittableBool isDirty;
+
+            // Bit masks for tracking which component properties were changed locally and need to be synced.
+            // Each byte tracks 8 component properties.
+<# for (var i = 0; i < (fieldDetailsList.Count / 8) + 1; i++) { #>
+            private byte dirtyBits<#= i #>;
+<# } #>
+
+            public bool IsDirty()
+            {
+                return isDirty;
+            }
+
+            /*
+            The propertyIndex arguments starts counting from 0. It depends on the order of which you defined
+            your component properties in a schema component but is not the schema field number itself. E.g.
+            component MyComponent
+            {
+                id = 1337;
+                bool val_a = 1;
+                bool val_b = 3;
+            }
+            In that case, val_a uses propertyIndex 0 and val_b uses propertyIndex 1 in this method.
+            */
+            public bool IsDirty(int propertyIndex)
+            {
+                if (propertyIndex < 0 || propertyIndex >= <#= fieldDetailsList.Count #>)
+                {
+                    throw new ArgumentException("propertyIndex argument out of range.");
+                }
+
+                var byteBatch = propertyIndex / 8;
```

Comment:
This looks suspicious - a comment explaining why it works would be good.
Doubly-suspicious is the need for the `default` case below. If the range checking logic is sound, that should be unnecessary.

Reply:
Removed the default case. That code should indeed never be called. Added a comment to clarify the approach.